expr/func_linearregression.go

```diff
@@ -0,0 +1,142 @@
+package expr
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/schema"
+	"github.com/raintank/dur"
+)
+
+type FuncLinearRegression struct {
+	in GraphiteFunc
+
+	startSourceAt string
+	endSourceAt   string
+
+	startTargetAt uint32
+	endTargetAt   uint32
+}
+
+func NewLinearRegression() GraphiteFunc {
+	return &FuncLinearRegression{}
+}
+
+func (s *FuncLinearRegression) Signature() ([]Arg, []Arg) {
+	return []Arg{
+			ArgSeriesList{
+				val: &s.in,
+			},
+			ArgString{
+				key:       "startSourceAt",
+				opt:       true,
+				validator: []Validator{IsATTime},
+				val:       &s.startSourceAt,
+			},
+			ArgString{
+				key:       "endSourceAt",
+				opt:       true,
+				validator: []Validator{IsATTime},
+				val:       &s.endSourceAt,
+			},
+		},
+		[]Arg{ArgSeriesList{}}
+}
+
+func (s *FuncLinearRegression) Context(context Context) Context {
+	s.startTargetAt = context.from
+	s.endTargetAt = context.to
+
+	return context
+}
+
+func linearRegressionAnalysis(series models.Series, startSourceAt uint32, endSourceAt uint32) (float64, float64, bool) {
+	var n float64
+	var sumI float64
+	var sumII float64
+	var sumV float64
+	var sumIV float64
+
+	for i := sort.Search(len(series.Datapoints), func(i int) bool { return series.Datapoints[i].Ts >= startSourceAt }); i < len(series.Datapoints) && series.Datapoints[i].Ts <= endSourceAt; i++ {
```

Comment:
Why do we need to do this? Why would we have more included datapoints than requested? How does this play with functions like `timeShift`

Reply:
This is just a product of me thinking that we used missing data points instead of NaN -- will fix.